components/feature/syncedtabs/src/main/java/mozilla/components/feature/syncedtabs/presenter/DefaultPresenter.kt

```diff
@@ -34,15 +33,15 @@ import mozilla.components.service.fxa.sync.SyncStatusObserver
  *
  */
 internal class DefaultPresenter(
-    private val context: Context,
     override val controller: SyncedTabsController,
     override val accountManager: FxaAccountManager,
     override val view: SyncedTabsView,
-    private val lifecycleOwner: LifecycleOwner
+    private val lifecycleOwner: LifecycleOwner,
+    private val enginesStorage: SyncEnginesStorage,
```

Comment:
I wish this was an interface instead of the concrete class.

Reply:
There isn't anything you can do here without feature creeping. I am just sighing. 🙂

Reply:
Do you mean that the `enginesStorage` were an interface? I feel your pain if so 😂 I'm a firm believer that all public library code should be an interface. I don't even like exposing concrete subtypes directly, I used to often do it indirectly through factory methods. Just naive stuff like:

```
interface Foo {
    companion object {
        fun build(p1, p2, p3): Foo = DefaultFoo(p1, p2, p3)
    }
}

private DefaultFoo(param1, param2, param3) : Foo
```

Still tied client implementations directly to a specific version of `build` (unless it didn't change when the concrete subtypes did), but it always felt like a good compromise between ease-of-use and the generic factory madness that things like `ViewModel` use.

Reply:
It doesn't feel like _too_ much scope creep to me to just change it to an interface, but we could also always do it separately in a pre-follow-up

Reply:
What would that API look like? 🤔 There is a balance of useful interfaces and just doing it for convention. I wanted to avoid feature creep here because I couldn't imagine myself what the API would look like to be useful (say, if we had multiple implementations for `SyncEnginesStorage`).

Reply:
I think it would be possible to create an API with the same surface that the current storage has, for example:

```
interface SyncEnginesStorage {
    fun getStatus(): Map<SyncEngine, Boolean>
    fun setStatus(engine: SyncEngine, status: Boolean)
    fun clear()
}
```

This would allow us to swap the concrete for the abstract type fairly easy. I generally agree it's feature creep, though, so I don't think there's necessarily a reason for doing it now. As for the matter of convention vs usefulness, that's why I think interfaces are especially important for library code. If that interface existed, it would allow us to do things like swap the underlying implementations based on a feature flag:

```
val enginesStorage: SyncEnginesStorage by lazy {
    if (FeatureFlags.syncEnginesV2) {
        DataStoreEnginesStorage()
    } else {
        SharedPreferencesEnginesStorage()
    }
}
```